File: examples/todo-app/spring-boot/src/main/java/com/example/todo/domain/TodoFilter.java
package com.example.todo.domain;

import org.jspecify.annotations.Nullable;

/** 一覧の絞り込み。クエリパラメータ filter=all|active|completed に対応する。 */
public enum TodoFilter {
    ALL,
    ACTIVE,
    COMPLETED;

    /** 不正な値や未指定は ALL として扱う。 */
    public static TodoFilter fromParam(@Nullable String value) {
        if (value == null) {
            return ALL;
        }
        for (TodoFilter filter : values()) {
            if (filter.name().equalsIgnoreCase(value)) {
                return filter;
            }
        }
        return ALL;
    }

    public String param() {
        return name().toLowerCase();
    }
}

File: examples/todo-app/rust/src/domain/todo_filter.rs
//! Translated from `TodoFilter` (TodoFilter.java) by Java2RustTranslator.

/// 一覧の絞り込み。クエリパラメータ filter=all|active|completed に対応する。
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash)]
pub enum TodoFilter {
    #[default]
    All,
    Active,
    Completed,
}

impl TodoFilter {
    /// すべての定数（`values()`）。
    pub fn values() -> [Self; 3] {
        [Self::All, Self::Active, Self::Completed]
    }

    /// 定数の名前（`name()`）。
    pub fn name(self) -> &'static str {
        match self {
            Self::All => "ALL",
            Self::Active => "ACTIVE",
            Self::Completed => "COMPLETED",
        }
    }

    /// 定数の位置（`ordinal()`）。
    pub fn ordinal(self) -> i32 {
        match self {
            Self::All => 0,
            Self::Active => 1,
            Self::Completed => 2,
        }
    }

    /// 名前の定数（`valueOf(name)`。なければ None）。
    pub fn value_of(name: &str) -> Option<Self> {
        match name {
            "ALL" => Some(Self::All),
            "ACTIVE" => Some(Self::Active),
            "COMPLETED" => Some(Self::Completed),
            _ => None,
        }
    }

    /// 不正な値や未指定は ALL として扱う。
    pub fn from_param(value: Option<&str>) -> TodoFilter {
        let Some(value) = value else {
            return TodoFilter::All;
        };
        for filter in TodoFilter::values() {
            if filter.name().to_lowercase() == value.to_lowercase() {
                return filter;
            }
        }
        TodoFilter::All
    }

    pub fn param(self) -> String {
        self.name().to_lowercase()
    }
}
